core/common/assets/js/api/core/data.js

```diff
@@ -0,0 +1,442 @@
+import Commands from './commands.js';
+import Cache from './data/cache';
+
+/**
+ * @typedef {('create'|'delete'|'get'|'update')} DataTypes
+ */
+
+/**
+ * @typedef {{}} RequestData
+ * @property {ComponentBase} component
+ * @property {string} command
+ * @property {string} endpoint
+ * @property {DataTypes} [type]
+ * @property {{}} [args]
+ * @property {number} [timestamp]
+ * @property {('hit'|'miss')} [cache]
+ */
+
+// TODO: Return it from the server. Original at WP_REST_Server.
+export const READABLE = [ 'GET' ],
+	CREATABLE = [ 'POST' ],
+	EDITABLE = [ 'POST', 'PUT', 'PATCH' ],
+	DELETABLE = [ 'DELETE' ],
+	ALLMETHODS = [ 'GET', 'POST', 'PUT', 'PATCH', 'DELETE' ];
+
+export default class Data extends Commands {
+	constructor( args = {} ) {
+		super( args );
+
+		this.args = Object.assign( args, {
+			namespace: 'elementor',
+			version: '1',
+		} );
+
+		this.cache = new Cache( this );
+
+		this.baseEndpointAddress = '';
+
+		elementorCommon.elements.$window.on( 'elementor:loaded', this.onElementorLoaded.bind( this ) );
+	}
+
+	onElementorLoaded() {
+		const { namespace, version } = this.args;
+
+		this.baseEndpointAddress = `${ elementor.config.rest_url }${ namespace }/v${ version }/`;
+	}
+
+	/**
+	 * Function getHTTPMethod().
+	 *
+	 * Returns HTTP Method by type.
+	 *
+	 * @param {DataTypes} type
+	 *
+	 * @return {string|boolean}
+	 */
+	getHTTPMethod( type ) {
+		switch ( type ) {
+			case 'create':
+				return 'POST';
+
+			case 'delete':
+				return 'DELETE';
+
+			case 'get':
+				return 'GET';
+
+			case 'update':
+				return 'PUT';
+		}
+
+		return false;
+	}
+
+	/**
+	 * Function getAllowedMethods().
+	 *
+	 * Returns allowed HTTP methods by type.
+	 *
+	 * @param {DataTypes} type
+	 *
+	 * @return {[string]|boolean}
+	 */
+	getAllowedMethods( type ) {
+		switch ( type ) {
+			case 'create':
+				return CREATABLE;
+
+			case 'delete':
+				return DELETABLE;
+
+			case 'get':
+				return READABLE;
+
+			case 'update':
+				return EDITABLE;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Function commandToEndpoint().
+	 *
+	 * Convert command to endpoint.
+	 *
+	 * For example `component/command/:arg_example` => `controller/endpoint/8`.
+	 *
+	 * TODO: Find a better solution.
+	 *
+	 * @param {string} command
+	 * @param {{}} args
+	 * @param {string|null} [format]
+	 *
+	 * @returns {string} endpoint
+	 */
+	commandToEndpoint( command, args, format = null ) {
```

Comment:
Can you provide a better explanation of the endpoint structure?

Reply:
please look here:
https://github.com/elementor/elementor/blob/31365f261c5a42fd7149b476acd8da30e9872e4f/tests/qunit/core/common/assets/js/api/core/data.spec.js#L37

Reply:
@iNewLegend Please change the placeholder format from `:id` to `{id}`.

Reply:
@matipojo, both php and js?

Reply:
Yes

Reply:
Handled. + tests added.